src/App.tsx

```diff
@@ -65,45 +70,44 @@ type Mgmt<T> =
  * if the state has actually changed.
  * 
  * @param actyx Actyx SDK instance
- * @param setter React state update function for the list of managed states
+ * @param map React state update function for the list of managed states
  * @param id ID of the plant or robot to be added
  * @param factory constructor for the function that will manage the state of this plant or robot
  */
-const add = <T extends Record<string, unknown>>(actyx: Actyx, setter: (_: (_: Mgmt<T>[]) => Mgmt<T>[]) => void, id: string,
-    factory: (_: Actyx, id: string, state: (_: T) => void, died: () => void) => ((() => void) | Promise<void>)) => {
+const add = <T extends Record<string, unknown>>(actyx: Actyx, sync: () => unknown, map: Map<string, Mgmt<T>>, id: string,
+    factory: (_: Actyx, id: string, state: (_: T) => void, died: () => void) => ((() => void))) => {
+  if (map.has(id)) return;
   console.log('adding', id)
 
   // start running or following a plant or robot
   const res = factory(actyx, id, (state) => {
-    setter((set) =>
-      deepEqual(state, set.find(x => x.id === id))
-        ? set : set.map(x => x.id === id ? { ...x, type: 'ready', state }
-        : x))
-  }, () => del(setter, id))
+    const old = map.get(id);
+    if (!old) return
+    const readyState = { ...old, type: 'ready' as const, state };
+    map.set(id, readyState);
+    sync()
+  }, () => del(sync, map, id))
+
+  const state = { type: 'fresh' as const, cancel: res, id }
 
-  const cancel = res instanceof Promise ? () => {} : res
-  const state = { type: 'fresh' as const, cancel, id }
+  map.set(id, state);
+  sync()
 
-  // ensure that the state is already in the list when the factory uses the setter
-  setter(set => [...set, state])
+  return res
 }
 
 /**
  * Remove a plant or robot from the list of managed states.
  * This submits the corresponding change logic to React, which will call the provided
  * function whenever it pleases (usually many times!).
  * 
- * @param setter React state update function for the list of managed states
+ * @param map React state update function for the list of managed states
  * @param id ID of the plant or robot to be removed
  */
-const del = <T extends Record<string, unknown>>(setter: (_: (_: Mgmt<T>[]) => Mgmt<T>[]) => void, id: string) => setter((set) => {
-  const idx = set.findIndex(x => x.id === id)
-  if (idx < 0) return set
-  console.log('removing', id)
-  const next = [...set]
-  next.splice(idx, 1)[0].cancel()
-  return next
-})
+const del = <T extends Record<string, unknown>>(sync: () => unknown, map: Map<string, Mgmt<T>>, id: string) => {
+  map.delete(id)
```

Comment:
this is not enough: when stopping to follow a plant or robot, the machine-runner needs to be destroyed right here

Reply:
ah I see, the missing `cancel` call